lookout/style/format/feature_extractor.py

```diff
@@ -0,0 +1,585 @@
+                        Position(offset, lineno, col),
+                        y=cls, path=path)
+                indentation = indentation[:len(line)]
+                if indentation:
+                    yield VirtualNode(
+                        "".join(indentation),
+                        Position(offset, lineno, col),
+                        node.end, path=path)
+            else:
+                if indentation:
+                    yield VirtualNode(
+                        "".join(indentation),
+                        Position(offset - len(line), lineno, col - len(line)),
+                        Position(offset - len(line) + len(indentation),
+                                 lineno,
+                                 col - len(line) + len(indentation)),
+                        path=path)
+                offset += - len(line) + len(indentation)
+                col += - len(line) + len(indentation)
+                if not my_indent:
+                    # indentation is the same
+                    continue
+                # indentation increase
+                for i, char in enumerate(my_indent):
+                    indentation.append(char)
+                    if char == "\t":
+                        cls = CLASS_INDEX[CLS_TAB_INC]
+                    else:
+                        cls = CLASS_INDEX[CLS_SPACE_INC]
+                    yield VirtualNode(
+                        char,
+                        Position(offset + i, lineno, col + i),
+                        Position(offset + i + 1, lineno, col + i + 1),
+                        y=cls, path=path)
+
+    @staticmethod
+    def _add_noops(vnodes: Sequence[VirtualNode], path: str) -> List[VirtualNode]:
+        """
+        Add CLS_NOOP nodes in between each node in the input sequence.
+
+        :param vnodes: The sequence of `VirtualNode`-s to augment with noop nodes.
+        :param path: path to file.
+        :return: The augmented `VirtualNode`-s sequence.
+        """
+        result = [VirtualNode(value="", start=Position(0, 1, 1), end=Position(0, 1, 1),
+                              y=CLASS_INDEX[CLS_NOOP], path=path)]
+        for vnode in vnodes:
+            result.append(vnode)
+            result.append(VirtualNode(value="", start=vnode.end, end=vnode.end,
+                                      y=CLASS_INDEX[CLS_NOOP], path=path))
+        return result
+
+    @staticmethod
+    def _index_nodes(vnodes: Sequence[VirtualNode]) -> None:
+        global_index = 0
+        labeled_index = 0
+        for vnode in vnodes:
+            vnode.global_index = global_index
+            global_index += 1
+            if vnode.y is not None:
+                vnode.labeled_index = labeled_index
+                labeled_index += 1
+
+    def _get_features(self, features: Sequence[str], sibling: Union[VirtualNode, bblfsh.Node],
+                      node: VirtualNode) -> List[int]:
+        for feature in sorted(features):
```

Comment:
Still not sure we should sort here.
Will we have a right order for feature names according to the columns of `X`?

Reply:
Thanks for reminding, I didn't have time to check that before the meetup :+1: